Validate tx

teadetime opened the pull request:
Draft PR for first round feedback

eito-fis commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,146 @@
+  mempool_tx = mempool_find(input->prev_tx_id);
+  if(mempool_tx != NULL){
+    return 1;
+  }
```
I also think this functionality is captured by checking if the UTXO is in the UTXO pool - this might be unnecessary? (Ex: if a tx is in the mempool, its outputs shouldn't be in the UTXO pool)

teadetime replied:
I think you are right on both counts, I'm going to remove this function

eito-fis commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,146 @@
+  if(memcmp(old_utxo->public_key_hash, test_pub_key_hash, PUB_KEY_HASH_LEN) != 0){
+    return 2;
+  }
+  return 0;
```
I think we were also planning on checking to see if the spent flag was set on the found UTXO, although I'm not sure if we want to follow through with that.

Desired functionality would be to mark a tx invalid if the UTXO it references was already spent by a tx in the mempool. This does raise some questions about user functionality, since it means you couldn't send a second tx that references a UTXO if you already sent one that spent it. Also raises issues during block validation time, since the tx in the block is likely (although not necessarily) also in the mempool.

IIRC the purpose of this feature was to avoid additional overhead when selecting things to mine - specifically, we need to make sure we don't have two transactions that reference the same UTXO. We would also need to clean the mempool of transactions that become invalid after one of their UTXOs are spent.

Given the issues outlined above, my vote would be to get rid of this in validation, and handle it in a new PR  down the line. This would have to add to block validation and block handling, and will probably be non-trivial (some design decision to make) and isn't critical for our demo, so I would also vote to do this later.

Curious to hear your thoughts.

eito-fis replied:
I think you bring up a good point in checking for double spend in block verification. We don't do it right now and we need to.

However, I disagree on requesting txs. We can't trust a node to send valid transactions, so we'll have to validate everything anyways. Moreover, we couldn't trust any data the other node sends that isn't already in the block so there isn't much point. It would be much more efficient to just check the block for double spend.

Here's my proposal:
- Update block verification to check for double spend
- Update mine transaction collection to avoid double spend
- Update handle block to clean mempool
- Remove spent flag from UTXOs (not wallet, which is separate)

This doesn't matter for the demo or this PR, so I would just say we open an issue and do it later.

(I will double check what bitcoin does when I get the chance)

eito-fis commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,146 @@
+  }
+  if(tx->outputs != NULL){
+    return 4;
+  }
```
Doesn't return 0

eito-fis replied:
This functionality also seems a bit off. Currently, if any input or output exists, we return a non-zero value. `validate_tx` handles a non-zero as an error, so this seems off.

I also think we should only throw an error if outputs don't exist. No inputs could be valid if it just contains  coinbase tx.

teadetime replied:
agreed!, intersting idea, will our serialization break if we have arrays set to NULL? if coinbase tx with num_inputs = 0 an inputs = NULL

COINBASE TX CURRENTLY SETS INPUTS=NULL no clue how we are serializing this but somehow block hashing is working...

eito-fis replied:
Doesn't seem to break anything, seems fine to me. If you wanna update the serialization to not Malloc when num is 0 that seems alright too. 

I'm alright leaving as is and not handle coinbase, but would prefer a comment pointing that out if so.

eito-fis reviewed (commented):
Structures is looking great, some functionality things to iron out. I think this is most of the way there though.

(Unrelated - we should consider replacing all the integer return values with proper error code constants at some point)

teadetime commented on src/core/txs/base_tx.c:
```diff
@@ -184,3 +184,9 @@ void print_tx(Transaction *tx, char *prefix){
 }
+
+void free_tx(Transaction *tx){
+  free(tx->inputs);
```
Noted that's a great catch!

teadetime commented on src/core/txs/sign_tx.c:
```diff
@@ -60,27 +60,57 @@ mbedtls_ecdsa_context *gen_keys() {
+void build_ctx_from_public(mbedtls_ecdsa_context *ctx, mbedtls_ecp_point *pub_key){
+  mbedtls_ecdsa_init(ctx);
+  mbedtls_ecp_group_load(&ctx->private_grp, CURVE);
+  ctx->private_Q = *pub_key;
```
Another good catch, will add comment to docstring

teadetime commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,146 @@
+  }
+  // need to make hash where the tx has signatures that are 0's
+  hash_tx(blank_hash, blank_sig_tx);
+  free_tx(blank_sig_tx);
```
copying keys seems like the right thing, alternatively we could set them to NULL after hashing and before freeing...less copying and same end result

eito-fis replied:
Any solution seems fine, up to you

teadetime commented:
Good point. I changed to allow Coinbase and added this to be used in the
Coinbase validation. The Coinbase validation is actually stricter and is
done separately. So I may make this strict and not allow empty inputs

On Wed, Mar 23, 2022, 5:46 PM Eamon Ito-Fisher ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/core/txs/validate_tx.c
> <https://github.com/olincollege/SoftSysOlinCoin/pull/20#discussion_r833749295>
> :
>
> > +  int input_unlockable = check_input_unlockable(input, blank_tx_hash);
> +}
> +
> +int validate_tx_not_null(Transaction *tx){
> +  if(tx->num_inputs != 0){
> +    return 1;
> +  }
> +  if(tx->inputs != NULL){
> +    return 2;
> +  }
> +  if(tx->num_outputs != 0){
> +    return 3;
> +  }
> +  if(tx->outputs != NULL){
> +    return 4;
> +  }
>
> Doesn't seem to break anything, seems fine to me. If you wanna update the
> serialization to not Malloc when num is 0 that seems alright too.
>
> I'm alright leaving as is and not handle coinbase, but would prefer a
> comment pointing that out if so.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/olincollege/SoftSysOlinCoin/pull/20#discussion_r833749295>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFOJTIVKHEK56NV7YZ23DHLVBOGMLANCNFSM5RJTE5FA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

eito-fis commented on src/tests/core/txs/test_validate_tx.c:
```diff
@@ -0,0 +1,289 @@
+  Transaction *tx1 = build_complex_tx();
+  for(unsigned int i = 0; i < tx1->num_inputs; i++){
+    mu_assert(
+      "Input Not unlockable",
```
Wrong error message

eito-fis commented on src/tests/core/txs/test_validate_tx.c:
```diff
@@ -0,0 +1,289 @@
+  last_key_pair = output_key_pair;
+  tx->num_inputs = 1;
+  tx->inputs = in;
+  tx->num_outputs = 1;
```
Would it be a significant amount of work to do 2 inputs and outputs? I think we might catch bugs that we otherwise won't. If it breaks our testing framework we've written so far then I don't think its worth it, but if its easy we should do it.

teadetime replied:
The hardest part is the signatures, I think we should do it. but after we do a mock, otherwise there will be a lot of weird key copying (or existing helper functions should return a set of keys that were generated to test signatures).

Alternatively, several tests can be run without signatures

eito-fis reviewed (commented):
Overall looking good! Thanks for your work so far on this. Some comments, nothing huge and only on thing which I think might cause runtime issues.

Re: Comments for tests. We talked offline about leaving the tests in a more draft state and refactoring once we get persistence. I think thats totally reasonable, and can ignore most of the test comments if you want.

teadetime commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,190 @@
+  if(tx == NULL){
+    return 1;
+  }
+  if(tx->num_inputs != 0){// Note 0 inputs happen for a coinbase tx and we expect inputs = Null
```
Agreed, this was edited before we realized we wanted separate functionality. Will look at the ouputs pattern

teadetime commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  for(unsigned int i = 0; i < tx->num_outputs; i++){
+    total_out += tx->outputs->amt;
```
Yeah that's a red flag, shows that we aren't testing bad transactions since they are hard to make(hopefully remedied soon)

teadetime commented on src/core/txs/validate_tx.c:
```diff
@@ -0,0 +1,190 @@
+
+  for(unsigned int i = 0; i < tx->num_outputs; i++){
+    total_out += tx->outputs->amt;
+    if(tx->outputs->public_key_hash == NULL){
```
Good catch, I guess the pub-key validation happens with looking if inputs can be unlocked etc

teadetime commented on src/tests/core/blocks/test_validate_block.c:
```diff
@@ -40,11 +45,17 @@ void _fill_mempool(){
   utxo_pool_add(input_tx, 0);
-
+  mbedtls_ecdsa_context *input_tx_context = malloc(sizeof(mbedtls_ecdsa_context));
+  memcpy(input_tx_context, last_key_pair, sizeof(mbedtls_ecdsa_context));
```
Interesting that you say this as I believe that the tests break without this. I agree testing should really just have proper data created/saved in a cleaner way

teadetime commented on src/includes/txs/base_tx.h:
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include <stdlib.h>
```
Must have gotten added in the mess of things...oops

teadetime commented on src/core/globals/double_spend_set.c:
```diff
@@ -0,0 +1,59 @@
+  set = NULL;
+}
+
+int double_spend_add(UTXOPool *set, unsigned char tx_id[TX_HASH_LEN], unsigned int vout) {
```
I don't think it changes much other than naming, and don't think the explicit tie to UTXO is actually too bad... Would rather prioritize other things.
if anything making something that was a generic set would be worthwhile
